File: src/leetcode/arrays_strings/1929_concatenation_of_array.rb
# frozen_string_literal: true

# LeetCode
# https://leetcode.com/problems/concatenation-of-array/description/

# Time complexity: O(n)
# Space complexity: O(n)

# @param {Integer[]} nums
# @return {Integer[]}
def get_concatenation(nums)
  length = nums.length

  double_length = 2 * length

  (0...double_length).map { |index| nums[index % length] }

  # Alternatives
  # Array.new(nums.length * 2) { |index| nums[index % nums.length] }
  # nums + nums
  # nums * 2
  # nums.concat(nums)
end

File: src/leetcode/arrays_strings/1929_concatenation_of_array.ts
// LeetCode
// https://leetcode.com/problems/concatenation-of-array/description/

// Time complexity: O(n)
// Space complexity: O(n)

const getConcatenation = (nums: number[]): number[] => {
  // In-built [...nums, ...nums] or .concat or use for loop

  const length = nums.length;

  const doubleLength = length * 2;

  const ans = new Array(doubleLength);

  nums.forEach((num, index) => {
    ans[index] = num;
    ans[length + index] = num;
  });

  return ans;
};

export default getConcatenation;
